### DISCUSSION TITLE
[Nuxt 3] Is there a hook that executed before `vueApp.use(router)`?

### DISCUSSION THREAD

**Discussion by yaquawa** [nuxt/nuxt#17978] [Questions]
I want to setup a global variable `isUser` that indicates the current user is whether a guest or a signed in user before vue-router started(before Nuxt's route middlewares be called), so I could guard routes where require authentication.

Is there a hook let me run my code before nuxt add the vue-router to the vue app ?

**Comment by fallsimply** [comment]
You can use a [global or named route middleware](https://v3.nuxtjs.org/docs/directory-structure/middleware) for this. 

To make this work, you could create named middleware and [use the middleware on the routes you need to guard using `definePageMeta`](https://v3.nuxtjs.org/docs/directory-structure/middleware/#example-a-named-route-middleware).

Additional Documentation: [Nuxt 3 - Middleware](https://v3.nuxtjs.org/docs/directory-structure/middleware)

**Comment by yaquawa** [comment]
Thanks @fallsimply , I understand I can use a middleware for auth guard, but my question is 

> Is there a hook that executed before `vueApp.use(router)`? 

not only for auth guard, I just run some code before `vueApp.use(router)`.

By the way, some pages show different UI based on the user whether a guest or a signed in user, so I don't want to guard  the whole route.

**Comment by viii47** [comment]
Hey @yaquawa, 

there's currently a lot of possibilities how to accomplish this. Personally I use a similar approach as @Diizzayy (See https://github.com/nuxt/framework/discussions/2701#discussioncomment-1961817). In my opinion the easiest way to ensure that both, SSR and onpage routing is properly checked without too much writeup. 

You can easily incorporate your desired method, for example reading the Cookies or have a callback function to verify the current user via your preferred method (token, cookie, login credentials, session id, ...)

**Reply by yaquawa** [reply]
Thanks @viii47 , but I'm not using SSR, I just want to know "Is there a hook that executed before `vueApp.use(router)`? "
